v2/spiffeid/trustdomain.go

```diff
@@ -5,42 +5,43 @@ import (
 	"strings"
 )
 
-// TrustDomain is the name of a SPIFFE trust domain (e.g. example.org).
+// TrustDomain represents the trust domain portion of a SPIFFE ID (e.g.
+// example.org).
 type TrustDomain struct {
 	name string
 }
 
 // TrustDomainFromString returns a new TrustDomain from a string. The string
-// can either be the host part of a URI authority component (e.g. example.org),
-// or a valid SPIFFE ID URI (e.g. spiffe://example.org), otherwise an error is
-// returned.  The trust domain is normalized to lower case.
-func TrustDomainFromString(s string) (TrustDomain, error) {
-	if !strings.Contains(s, "://") {
-		s = "spiffe://" + s
+// can either be a trust domain name (e.g. example.org), or a valid SPIFFE ID
+// URI (e.g. spiffe://example.org), otherwise an error is returned.
+// See https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#21-trust-domain.
+func TrustDomainFromString(idOrName string) (TrustDomain, error) {
+	// Something looks kinda like a scheme separator, let's try to parse as
+	// an ID. We use :/ instead of :// since the diagnostics are better for
+	// a bad input like spiffe:/trustdomain.
+	if strings.Contains(idOrName, ":/") {
+		id, err := FromString(idOrName)
+		if err != nil {
+			return TrustDomain{}, err
+		}
+		return id.TrustDomain(), nil
 	}
 
-	id, err := FromString(s)
-	if err != nil {
-		return TrustDomain{}, err
+	if idOrName == "" {
```

Comment:
nit: it probably makes more sense to check this condition first since `idOrName` will never contain `:/` when it is `""`

Reply:
👍